src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,279 @@
+            StorageItem item = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Record).Add(hash_domain).Add(GetKey(Utility.StrictUTF8.GetBytes(name))).Add(type), () => new StorageItem());
+            item.Value = Utility.StrictUTF8.GetBytes(data);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.ReadStates)]
+        public string GetRecord(StoreView snapshot, string name, RecordType type)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string domain = string.Join('.', name.Split('.')[^2..]);
+            byte[] hash_domain = GetKey(Utility.StrictUTF8.GetBytes(domain));
+            StorageItem item = snapshot.Storages.TryGet(CreateStorageKey(Prefix_Record).Add(hash_domain).Add(GetKey(Utility.StrictUTF8.GetBytes(name))).Add(type));
+            if (item is null) return null;
+            return Utility.StrictUTF8.GetString(item.Value);
+        }
+
+        public IEnumerable<(RecordType Type, string Data)> GetRecords(StoreView snapshot, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string domain = string.Join('.', name.Split('.')[^2..]);
+            byte[] hash_domain = GetKey(Utility.StrictUTF8.GetBytes(domain));
+            foreach (var (key, value) in snapshot.Storages.Find(CreateStorageKey(Prefix_Record).Add(hash_domain).Add(GetKey(Utility.StrictUTF8.GetBytes(name))).ToArray()))
+                yield return ((RecordType)key.Key[^1], Utility.StrictUTF8.GetString(value.Value));
+        }
+
+        [ContractMethod(0_01000000, CallFlags.WriteStates)]
+        private void DeleteRecord(ApplicationEngine engine, string name, RecordType type)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string domain = string.Join('.', name.Split('.')[^2..]);
+            byte[] hash_domain = GetKey(Utility.StrictUTF8.GetBytes(domain));
+            NameState state = engine.Snapshot.Storages[CreateStorageKey(Prefix_Token).Add(hash_domain)].GetInteroperable<NameState>();
+            if (!engine.CheckWitnessInternal(state.Owner) || !engine.CheckWitnessInternal(state.Admin)) throw new InvalidOperationException();
+            engine.Snapshot.Storages.Delete(CreateStorageKey(Prefix_Record).Add(hash_domain).Add(GetKey(Utility.StrictUTF8.GetBytes(name))).Add(type));
+        }
+
+        [ContractMethod(0_03000000, CallFlags.ReadStates)]
+        public string Resolve(StoreView snapshot, string name, RecordType type)
+        {
+            return Resolve(snapshot, name, type, 2);
+        }
+
+        private string Resolve(StoreView snapshot, string name, RecordType type, int redirect)
+        {
+            if (redirect < 0) throw new InvalidOperationException();
+            var dictionary = GetRecords(snapshot, name).ToDictionary(p => p.Type, p => p.Data);
```

Comment:
I think that we can avoid the dictionary with `.Where`

Reply:
No much difference.